### PR TITLE
Add test coverage for the Network Model

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

File: firmware/src/network_model.c [hunk 1]
```diff
@@ -147,8 +155,8 @@ static int LookupIndex(unsigned int id) {
  *
  * This randomly fails and returns 0.0.0.0.
  */
-static uint32_t GetDHCPAddress() {
-  if (Random_PseudoGet() % DHCP_FAILURE_RATIO == 0) {
+static uint32_t GetDHCPAddress(bool can_fail) {
+  if (can_fail && Random_PseudoGet() % DHCP_FAILURE_RATIO == 0) {
```
Comment (peternewman) [inline firmware/src/network_model.c]: Presumably flakeyness isn't a good thing, and testing flakeyness doesn't work too well either?

File: firmware/src/network_model.c [hunk 2]
```diff
@@ -181,7 +190,8 @@ static void ConfigureInterface(InterfaceState *interface) {
     interface->current_netmask = interface->configured_netmask;
     interface->config_source = CONFIG_SOURCE_STATIC;
   } else if (interface->configured_dhcp_mode &&
-             (dhcp_address = GetDHCPAddress())) {
+             (dhcp_address =
```
Comment (peternewman) [inline firmware/src/network_model.c]: I'm not personally a fan of this style of overloading, but I guess there isn't another sensible way to do it.

File: firmware/src/network_model.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/network_model.c]: Yup

File: tests/tests/NetworkModelTest.cpp [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+  };
+
+  unique_ptr<RDMResponse> response(GetResponseFromData(
+        request.get(), expected_response, arraysize(expected_response)));
+
+  int size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+
+  // test the NR_DATA_OUT_OF_RANGE
+  interface_id = HostToNetwork(2);
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Is it worth doing this as NETWORK_MODEL_COUNT (Or whatever it's called) + 1?

File: tests/tests/NetworkModelTest.cpp [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+  request = BuildGetRequest(PID_INTERFACE_LABEL,
+      reinterpret_cast<uint8_t*>(&interface_id), sizeof(interface_id));
+
+  response.reset(NackWithReason(request.get(), ola::rdm::NR_DATA_OUT_OF_RANGE));
+  size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+}
+
+TEST_F(NetworkModelTest, getHardwareAddress) {
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Test on one that's not type1 too.

File: tests/tests/NetworkModelTest.cpp (hunk 3 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: NUMBER_OF_INTERFACES  ? That's internal to the .c file

File: tests/tests/NetworkModelTest.cpp [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    0x00, 0x00, 0x00, 0x01, 0x00
+  };
+
+  unique_ptr<RDMResponse> response(GetResponseFromData(
+        request.get(), expected_response, arraysize(expected_response)));
+
+  int size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+
+  // Get the DHCP mode for the second (4th) interface.
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Pardon?

Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Are they not sequential?

Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: Correct. This is designed to trip people up :)

File: tests/tests/NetworkModelTest.cpp (hunk 4 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: Done

File: tests/tests/NetworkModelTest.cpp (hunk 3 above)
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Ah okay.

File: tests/tests/NetworkModelTest.cpp [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+    0x52, 0x12, 0x34, 0x56, 0x78, 0x9a
+  };
+
+  unique_ptr<RDMResponse> response(GetResponseFromData(
+        request.get(), expected_response, arraysize(expected_response)));
+
+  int size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+
+  // test the NR_DATA_OUT_OF_RANGE
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Clarify which mode

File: tests/tests/NetworkModelTest.cpp [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+  interface_id = HostToNetwork(3);
+  request = BuildGetRequest(PID_INTERFACE_HARDWARE_ADDRESS_TYPE1,
+      reinterpret_cast<uint8_t*>(&interface_id), sizeof(interface_id));
+
+  response.reset(NackWithReason(request.get(), ola::rdm::NR_DATA_OUT_OF_RANGE));
+  size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+
+  // test the NR_DATA_OUT_OF_RANGE
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Again

File: tests/tests/NetworkModelTest.cpp (hunk 5 above)
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Perhaps change it to id 4, rather than 4th.

File: tests/tests/NetworkModelTest.cpp [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+    0x00, 0x00, 0x00, 0x01, 0x00
+  };
+
+  unique_ptr<RDMResponse> response(GetResponseFromData(
+        request.get(), expected_response, arraysize(expected_response)));
+
+  int size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+
+  // Get the zeroconf mode for the second (4th) interface.
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Again

File: tests/tests/NetworkModelTest.cpp [hunk 9]
```diff
@@ -0,0 +1,485 @@
...
+  const uint8_t set_param_data[] = { 0x00, 0x00, 0x00, 0x01, 0x01 };
+
+  request = BuildSetRequest(PID_IPV4_DHCP_MODE, set_param_data,
+                            arraysize(set_param_data));
+  response.reset(GetResponseFromData(request.get()));
+
+  size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+
+  // And read it back to confirm
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Also check ipv4_current_address hasn't updated here.

File: tests/tests/NetworkModelTest.cpp [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+  };
+
+  unique_ptr<RDMRequest> request = BuildSetRequest(
+      PID_IPV4_STATIC_ADDRESS, param_data, arraysize(param_data));
+  unique_ptr<RDMResponse> response(GetResponseFromData(request.get()));
+
+  int size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+
+  // Read it back
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Test ipv4_current_address is still the original here.

File: tests/tests/NetworkModelTest.cpp [hunk 11]
```diff
@@ -0,0 +1,485 @@
...
+  index = 3;
+  request = BuildGetRequest(PID_DNS_NAME_SERVER, &index, sizeof(uint8_t));
+  response.reset(NackWithReason(request.get(), ola::rdm::NR_DATA_OUT_OF_RANGE));
+
+  size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+}
+
+TEST_F(NetworkModelTest, hostname) {
+  // Set the hostname
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Also set an oversize one.

File: tests/tests/NetworkModelTest.cpp [hunk 12]
```diff
@@ -0,0 +1,485 @@
...
+        request.get(), reinterpret_cast<const uint8_t*>(hostname),
+        strlen(hostname)));
+
+  size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+}
+
+TEST_F(NetworkModelTest, domainName) {
+  // Set the domain name
+  const char domain_name[] = "myco.co.nz";
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: A nice touch.

File: tests/tests/NetworkModelTest.cpp [hunk 13]
```diff
@@ -0,0 +1,485 @@
...
+  response.reset(GetResponseFromData(
+        request.get(), reinterpret_cast<const uint8_t*>(hostname),
+        strlen(hostname)));
+
+  size = InvokeRDMHandler(request.get());
+  EXPECT_THAT(ArrayTuple(g_rdm_buffer, size), ResponseIs(response.get()));
+}
+
+TEST_F(NetworkModelTest, domainName) {
+  // Set the domain name
```
Comment (peternewman) [inline tests/tests/NetworkModelTest.cpp]: Again try an oversize one.

**Comment by peternewman** [comment]
LGTM, just address the comments.

File: tests/tests/NetworkModelTest.cpp (hunk 6 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: Done

File: tests/tests/NetworkModelTest.cpp (hunk 7 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: Done

File: tests/tests/NetworkModelTest.cpp (hunk 5 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: Done

File: tests/tests/NetworkModelTest.cpp (hunk 9 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: These tests aren't supposed to be exhaustive, that sort of thing is best done in the responder tests.

File: tests/tests/NetworkModelTest.cpp (hunk 8 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: Done

File: tests/tests/NetworkModelTest.cpp (hunk 10 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: Same comment

File: tests/tests/NetworkModelTest.cpp (hunk 11 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: And same here. Covered by the responder tests

File: tests/tests/NetworkModelTest.cpp (hunk 13 above)
Comment (nomis52) [inline tests/tests/NetworkModelTest.cpp]: Same

**Comment by peternewman** [comment]
LGTM

**Pull request merged**